File: src/main/java/com/chattriggers/ctjs/triggers/OnTrigger.java
//package com.chattriggers.ctjs.triggers;
//
//import com.chattriggers.ctjs.engine.ILoader;
//import com.chattriggers.ctjs.engine.module.Module;
//import jdk.nashorn.internal.objects.Global;
//
//public abstract class OnTrigger {
//    protected Object method;
//    protected Priority priority;
//    protected TriggerType type;
//    protected Module owningModule;
//    protected ILoader loader;
//
//    public Module getOwningModule() {
//        return owningModule;
//    }
//
//    public void setOwningModule(Module owningModule) {
//        this.owningModule = owningModule;
//    }
//
//    public Object getMethod() {
//        return method;
//    }
//
//    public Priority getPriority() {
//        return priority;
//    }
//
//    public TriggerType getType() {
//        return type;
//    }
//
//    private Global global;
//
//    protected OnTrigger(Object method, TriggerType type, ILoader loader) {
//        this.method = method;
//        this.priority = Priority.NORMAL;
//        this.type = type;
//        this.loader = loader;
//
//        this.register();
//    }
//
//    /**
//     * Sets a triggers priority using {@link Priority}.
//     * Highest runs first.
//     * @param priority the priority of the trigger
//     * @return the trigger for method chaining
//     */
//    public OnTrigger setPriority(Priority priority) {
//        this.priority = priority;
//        return this;
//    }
//
//    /**
//     * Registers a trigger based on its type.
//     * This is done automatically with TriggerRegister.
//     * @return the trigger for method chaining
//     */
//    public OnTrigger register() {
//        this.loader.addTrigger(this);
//        return this;
//    }
//
//    /**
//     * Unregisters a trigger.
//     * @return the trigger for method chaining
//     */
//    public OnTrigger unregister() {
//        this.loader.removeTrigger(this);
//        return this;
//    }
//
//    /**
//     * @return boolean of if trigger is registered
//     */
//    public boolean isRegistered() {
//        return true;
//    }
//
//    protected void callMethod(Object... args) {
//        this.loader.trigger(this, this.method, args);
//    }
//
//    public abstract void trigger(Object... args);
//
//    public enum TriggerResult {
//        CANCEL
//    }
//
//    public enum Priority {
//        //LOWEST IS RAN LAST
//        LOWEST, LOW, NORMAL, HIGH, HIGHEST
//    }
//}

File: src/main/kotlin/com/chattriggers/ctjs/triggers/OnTrigger.kt
package com.chattriggers.ctjs.triggers

import com.chattriggers.ctjs.Reference
import com.chattriggers.ctjs.engine.ILoader
import com.chattriggers.ctjs.utils.kotlin.External

@External
abstract class OnTrigger protected constructor(
    var method: Any,
    var type: TriggerType,
    protected var loader: ILoader
) : Comparable<OnTrigger> {
    var priority: Priority = Priority.NORMAL
        private set

    init {
        register()
    }

    /**
     * Sets a triggers priority using [Priority].
     * Highest runs first.
     * @param priority the priority of the trigger
     * @return the trigger for method chaining
     */
    fun setPriority(priority: Priority): OnTrigger {
        this.priority = priority
        return this
    }

    /**
     * Registers a trigger based on its type.
     * This is done automatically with TriggerRegister.
     * @return the trigger for method chaining
     */
    fun register() = apply {
        this.loader.addTrigger(this)
    }

    /**
     * Unregisters a trigger.
     * @return the trigger for method chaining
     */
    fun unregister() = apply {
        this.loader.removeTrigger(this)
    }

    protected fun callMethod(args: Array<out Any?>) {
        if (!Reference.isLoaded) return

        this.loader.trigger(this, this.method, args)
    }

    abstract fun trigger(args: Array<out Any?>)

    override fun compareTo(other: OnTrigger): Int {
        val ordCmp = priority.ordinal - other.priority.ordinal
        return if (ordCmp == 0)
            hashCode() - other.hashCode()
        else ordCmp
    }

    enum class Priority {
        //LOWEST IS RAN LAST
        HIGHEST,
        HIGH, NORMAL, LOW, LOWEST
    }
}
